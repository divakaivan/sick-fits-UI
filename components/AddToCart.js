import React from "react";
import {Mutation} from "react-apollo";
import gql from "graphql-tag";
import {CURRENT_USER_QUERY} from "./User";

export const ADD_TO_CART_MUTATION = gql`
    mutation addToCart($id: ID!) {
        addToCart(id: $id) {
            id
            quantity
        }
    }
`;

class AddToCart extends React.Component {

    render() {
        const {id} = this.props;

        return(
            <Mutation refetchQueries={[{query: CURRENT_USER_QUERY}]} variables={{id}} mutation={ADD_TO_CART_MUTATION}>
                {(addToCart, {loading})=>(
                    <button disabled={loading} onClick={addToCart}>Add{loading && 'ing'} to cart</button>
                )}
            </Mutation>
        )
    }
}

export default AddToCart;