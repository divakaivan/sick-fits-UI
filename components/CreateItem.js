import React, {Component} from 'react';
import {Mutation} from "react-apollo";
import gql from "graphql-tag";
import Router from 'next/router';

import Form from "./styles/Form";
import formatMoney from "../lib/formatMoney";
import Error from "./ErrorMessage";

const CREATE_ITEM_MUTATION = gql`
    mutation CREATE_ITEM_MUTATION(
    $title: String!
    $description: String!
    $price: Int!
    $image: String
    $largeImage: String
    ) {
    createItem(
    title: $title
    description: $description
    price: $price
    image: $image
    largeImage: $largeImage
    ) {id}
    }
`;

class CreateItem extends Component {
    state = {
        title: '',
        description: '',
        image: '',
        largeImage: '',
        price: 0
    };

     handleChange = e => {
        const {name, type, value} = e.target;
        const val = type === 'number' ? parseFloat(value) : value;
        this.setState({[name]: val})
    };

    uploadFile = async e => {
        const files = e.target.files;
        const data = new FormData();
        data.append('file', files[0]);
        data.append('upload_preset', 'sick-fits');

        const res = await fetch('https://api.cloudinary.com/v1_1/dsiqrsfbi/image/upload',
            {method: 'POST', body: data});
        const file = await res.json();

        this.setState({
            image: file.secure_url,
            largeImage: file.eager[0].secure_url
        });
    };

    render() {
        return (
            <Mutation
                mutation={CREATE_ITEM_MUTATION}
                variables={this.state}
            >
                {(createItem, {loading, error, called, data})=>(
                    <Form onSubmit={async e=>{
                        // stop the form from submitting
                        e.preventDefault();
                        // call the mutation
                        const res = await createItem();
                        // change them to the single item page
                        console.log(res);
                        Router.push({
                            pathname: '/item',
                            query: {id: res.data.createItem.id}
                        });
                    }}>
                        <Error error={error}/>
                        <fieldset disabled={loading} aria-busy={loading}>
                            <label htmlFor="file">
                                Image
                                <input
                                    type="file"
                                    id="file"
                                    name="file"
                                    placeholder="Upload an image"
                                    onChange={this.uploadFile}
                                    required/>
                                {this.state.image && <img src={this.state.image} alt="Upload preview"/>}
                            </label>

                            <label htmlFor="title">
                                Title
                                <input
                                    value={this.state.title}
                                    type="text"
                                    id="title"
                                    name="title"
                                    placeholder="Title"
                                    onChange={this.handleChange}
                                    required/>
                            </label>

                            <label htmlFor="price">
                                Price
                                <input
                                    value={this.state.price}
                                    type="number"
                                    id="price"
                                    name="price"
                                    placeholder="Price"
                                    onChange={this.handleChange}
                                    required/>
                            </label>

                            <label htmlFor="description">
                                Description
                                <textarea
                                    value={this.state.description}
                                    id="description"
                                    name="description"
                                    placeholder="Enter a description"
                                    onChange={this.handleChange}
                                    required/>
                            </label>
                            <button type="submit">Submit</button>
                        </fieldset>
                    </Form>
                )}
            </Mutation>

        );
    }
}

export default CreateItem;
export {CREATE_ITEM_MUTATION};
