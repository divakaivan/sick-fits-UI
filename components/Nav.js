import Link from "next/link";
import React from "react";

import User from "./User";
import NavStyles from "./styles/NavStyles";
import Signout from "./Signout";
import {Mutation} from "react-apollo";
import {TOGGLE_CART_MUTATION} from "./Cart";
import CartCount from "./CartCount";

const Nav = () => (
    <User>
        {({data: {me}}) => (
            <NavStyles data-test="nav">
                <Link href="/items"><a>Shop</a></Link>
                {me && (
                    <> {/* because we use next.js which uses babel7, this is the same as React.Fragment */}
                        <Link href="/sell"><a>Sell</a></Link>
                        <Link href="/orders"><a>Orders</a></Link>
                        <Link href="/me"><a>Account</a></Link>
                        <Signout/>
                        <Mutation mutation={TOGGLE_CART_MUTATION}>
                            {toggleCart=>(
                                <button onClick={toggleCart}>My cart <CartCount
                                    count={me.cart.reduce((tally, cartItem) => tally + cartItem.quantity, 0)}/></button>
                            )}
                        </Mutation>
                    </>
                )}
                {!me && (
                    <Link href="/signup"><a>Sign in</a></Link>
                )}
            </NavStyles>
        )}
    </User>

);

export default Nav;