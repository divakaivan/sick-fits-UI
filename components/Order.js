import React, {Component} from 'react';
import PropTypes from 'prop-types';
import {Query} from "react-apollo";
import {format} from "date-fns";
import Head from "next/head";
import gql from "graphql-tag";
import formatMoney from "../lib/formatMoney";
import Error from "./ErrorMessage";
import OrderStyles from "./styles/OrderStyles";


export const SINGLE_ORDER_QUERY = gql`
    query SINGLE_ORDER_QUERY($id: ID!) {
        order(id: $id) {
            id
            charge
            total
            createdAt
            user {
                id
            }
            items {
                id
                title
                description
                price
                image
                quantity
            }
        }
    }
`;

class Order extends Component {
    render() {
        return (
            <Query query={SINGLE_ORDER_QUERY} variables={{id: this.props.id}}>
                {({data, error, loading})=> {
                    if (error) return <Error error={error}/>;
                    if (loading) return <p>Loading...</p>;
                    const order = data.order;
                    return (
                        <OrderStyles data-test="order">
                            <Head><title>Sick Fits - Order {order.id}</title></Head>
                            <p>
                                <span>Order ID:</span>
                                <span>{this.props.id}</span>
                            </p>
                            <p>
                                <span>Charge</span>
                                <span>{order.charge}</span>
                            </p>
                            <p>
                                <span>Date</span>
                                <span>{format(order.createdAt, 'dd MMM, YYYY h:mm a')}</span>
                            </p>
                            <p>
                                <span>Order Total</span>
                                <span>{formatMoney(order.total)}</span>
                            </p>
                            <p>
                                <span>Item Count</span>
                                <span>{order.items.length}</span>
                            </p>
                            <div className="items">
                                {order.items.map(item=>(
                                    <div key={item.id} className="order-item">
                                        <img src={item.image} alt={item.title}/>
                                        <div className="item-details">
                                            <h2>{item.title}</h2>
                                            <p>Qty: {item.quantity}</p>
                                            <p>Each: {formatMoney(item.price)}</p>
                                            <p>SubTotal: {formatMoney(item.price * item.quantity)}</p>
                                            <p>{item.description}</p>
                                        </div>
                                    </div>
                                ))}
                            </div>
                        </OrderStyles>
                    )
                }}
            </Query>
        );
    }
}

Order.propTypes = {
    id: PropTypes.string.isRequired
};

export default Order;