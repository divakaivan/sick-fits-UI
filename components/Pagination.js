import React from "react";
import gql from "graphql-tag";
import {Query} from "react-apollo";
import Head from "next/head";
import Link from "next/link";

import PaginationStyles from "./styles/PaginationStyles";
import {perPage} from "../config";

export const PAGINATION_QUERY = gql`
    query PAGINATION_QUERY {
        # in the db this accepts a where input, but because we just need the items, 
        # aka we dont need to categorize or filter the items, we dont need to pass in anything
        itemsConnection {
            aggregate {
                count
            }
        }
    }
`;

const Pagination = props => (
    <Query query={PAGINATION_QUERY}>
        {({data, loading, error}) => {
            if (loading) return <p>Loading...</p>;
            const {count} = data.itemsConnection.aggregate;
            const pages = Math.ceil(count / perPage);
            return (
                <PaginationStyles data-test="pagination">
                    <Head>
                        <title>Sick Fits! Page {props.page} of {pages || 1}</title>
                    </Head>
                    <Link prefetch href={{
                        pathname: 'items',
                        query: {page: props.page - 1}
                    }}>
                        <a className="prev" aria-disabled={props.page <= 1}>{"<"} Prev</a>
                    </Link>
                    <p>Page {props.page} of <span className="totalPages">{pages || 1}</span></p>
                    <p>{count} Items total</p>
                    <Link prefetch href={{
                        pathname: 'items',
                        query: {page: props.page + 1}
                    }}>
                        <a className="next" aria-disabled={props.page >= pages}>Next {">"}</a>
                    </Link>
                </PaginationStyles>

            )
        }}
    </Query>
);

export default Pagination