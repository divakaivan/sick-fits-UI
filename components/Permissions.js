import React from "react";
import {Query} from "react-apollo";
import Error from "./ErrorMessage";
import gql from "graphql-tag";

import Table from "./styles/Table";
import SickButton from "./styles/SickButton";

const possiblePermissions = [
    "ADMIN",
    "USER",
    "ITEMCREATE",
    "ITEMUPDATE",
    "ITEMDELETE",
    "PERMISSIONUPDATE"
];

const ALL_USERS_QUERY = gql`
    query {
        users{
            id
            name
            email
            permissions
        }
    }
`;

const Permissions = props => (
    <Query query={ALL_USERS_QUERY}>
        {({data, loading, error}) => (
            <div>
                <Error error={error}/>
                <div>
                    <h2>Manage permissions</h2>
                    <Table>
                        <thead>
                        <tr>
                            <th>name</th>
                            <th>Email</th>
                            {possiblePermissions.map((permission, i)=><th key={i}>{permission}</th>)}
                            <th>\/</th>
                        </tr>
                        </thead>
                        <tbody>
                        {data.users.map((user, i) => <User key={i} user={user}/>)}
                        </tbody>
                    </Table>
                </div>
            </div>
        )}
    </Query>
);

class User extends React.Component {
    render() {
        const user = this.props.user;
        return (
            <tr>
                <td>{user.name}</td>
                <td>{user.email}</td>
                {possiblePermissions.map((permission, i)=>(
                    <td key={i}>
                        <label htmlFor={`${user.id}-permission-${permission}`}>
                            <input type="checkbox"/>
                        </label>
                    </td>
                ))}
                <td>
                    <SickButton>Update</SickButton>
                </td>
            </tr>
        )
    }
}

export default Permissions;
