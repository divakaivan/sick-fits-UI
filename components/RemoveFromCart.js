import React from "react";
import {Mutation} from "react-apollo";
import styled from "styled-components";
import PropTypes from "prop-types";
import gql from "graphql-tag";

import {CURRENT_USER_QUERY} from "./User";

export const REMOVE_FROM_CART_MUTATION = gql`
    mutation removeFromCart($id: ID!) {
        removeFromCart(id: $id) {
            id
        }
    }
`;

const BigButton = styled.button`
 font-size: 3rem;
 background: none;
 border: 0;
 &:hover {
  color: ${props => props.theme.red};
  cursor: pointer;
 }
`;

class RemoveFromCart extends React.Component {
    static propTypes = {
        id: PropTypes.string.isRequired
    };

    // this gets called as soon as we get a response back from the server after a mutation
    // has been performed
    update = (cache, payload) => {
        // 1. read the cache
        const data = cache.readQuery({query: CURRENT_USER_QUERY});
        // 2. remove that item from the cart
        const cartItemId = payload.data.removeFromCart.id;
        data.me.cart = data.me.cart.filter(cartItem => cartItem.id !== cartItemId)
        // 3. write it back to the cache
        cache.writeQuery({query: CURRENT_USER_QUERY, data})
    };

    render() {
        return (
            <Mutation
                variables={{id: this.props.id}}
                mutation={REMOVE_FROM_CART_MUTATION}
                optimisticResponse={{
                    __typename: 'Mutation',
                    removeFromCart: {
                        __typename: 'CartItem',
                        id: this.props.id
                    }
                }} // this is what is returned immediately from the backend, so the delete operation is faster.
                // we fake the successful return because we assume it will be successful, and update the UI immediately,
                // but the backend still takes 1s or so to actually delete the item from the cart
                update={this.update}
            >
                {(removeFromCart, {loading, error}) => (
                    <BigButton onClick={() => {
                        removeFromCart().catch(err => alert(err.message));

                    }} disabled={loading} title="Delete Item">
                        &times;
                    </BigButton>
                )}
            </Mutation>
        );
    }
}

export default RemoveFromCart;