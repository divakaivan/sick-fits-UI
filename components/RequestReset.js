import React, {Component} from 'react';
import {Mutation} from "react-apollo";
import gql from "graphql-tag";
import Form from "./styles/Form";
import Error from "./ErrorMessage";


export const REQUEST_RESET_MUTATION = gql`
    mutation REQUEST_RESET_MUTATION($email: String!) {
        requestReset(email: $email) {
            message
        }
    }
`;

class RequestReset extends Component {
    state = {
        email: ''
    };

    saveToState = e => {
        this.setState({
            [e.target.name]: e.target.value
        })
    };

    render() {
        return (
            <Mutation mutation={REQUEST_RESET_MUTATION} variables={this.state}>
                {(reset, {error, loading, called})=>{
                    return <Form method="post" data-test="form" onSubmit={async (e)=>{
                        e.preventDefault();
                        await reset();
                        this.setState({email: ''});
                    }
                    }>
                        {/* we need method=post, because by default form elements are GET and after submit,
                        it will refresh the page and put the submitted info in the url */}
                        <fieldset disabled={loading} aria-busy={loading}>
                            <h2>Request a password reset</h2>
                            <Error error={error}/>
                            {!error && !loading && called && <p>Success! Check your email for a reset link!</p>}
                            <label htmlFor="email">
                                Email
                                <input
                                    type="email"
                                    name="email"
                                    placeholder="Enter your email"
                                    value={this.state.email}
                                    onChange={this.saveToState}/>
                            </label>
                            <button type="submit">Request reset</button>
                        </fieldset>
                    </Form>
                }}
            </Mutation>
        );
    }
}

export default RequestReset;