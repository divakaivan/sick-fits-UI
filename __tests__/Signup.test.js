import React from "react";
import {mount} from "enzyme";
import toJson from "enzyme-to-json";
import {MockedProvider} from "react-apollo/test-utils";
import wait from "waait";
import {ApolloConsumer} from "react-apollo";
import Signup, {SIGNUP_MUTATION} from "../components/Signup";
import {CURRENT_USER_QUERY} from "../components/User";
import {fakeUser} from "../lib/testUtils";


function type(wrapper, name, value) {
    wrapper.find(`input[name="${name}"]`).simulate('change', {target: {name, value}})
}


const me = fakeUser();
const mocks = [
    // signup mock mutation
    {
        request: {query: SIGNUP_MUTATION, variables: {email: me.email, name: me.name, password: 'my-pass'}},
        result: {
            data: {
                signup: {
                    id: 'abc123',
                    ...me
                }
            }
        }
    },
    // current user query mock
    {
        request: {query: CURRENT_USER_QUERY},
        result: {
            data: {me}
        }
    }
];


describe('<Signup/>', () => {
    it('renders and matches snapshot', async () => {
        const wrapper = mount(
            <MockedProvider>
                <Signup/>
            </MockedProvider>
        );
        expect(toJson(wrapper.find('form'))).toMatchSnapshot();
    });

    it('calls the mutation properly', async () => {
        let apolloClient;
        const wrapper = mount(
            <MockedProvider mocks={mocks}>
                <ApolloConsumer>
                    {client=>{
                        apolloClient = client;
                        return <Signup/>
                    }}
                </ApolloConsumer>
            </MockedProvider>
        );
        await wait();
        wrapper.update();
        type(wrapper, 'name', me.name);
        type(wrapper, 'email', me.email);
        type(wrapper, 'password', 'my-pass');
        wrapper.update();
        wrapper.find('form').simulate('submit');
        await wait();
        // query the user out of the apollo client
        // because once the signup mutation is finished.
        // the user should be logged in because we refetch the current user query (Signup.js)
        const user = await apolloClient.query({query: CURRENT_USER_QUERY});
        expect(user.data.me).toMatchObject(me);
    });
});
