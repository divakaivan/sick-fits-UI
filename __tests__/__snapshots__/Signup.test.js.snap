// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Signup/> renders and matches snapshot 1`] = `
<form
  className="Form-sc-1xszr8q-0 krRxky"
  method="post"
  onSubmit={[Function]}
>
  <fieldset
    aria-busy={false}
    disabled={false}
  >
    <h2>
      Sign Up for an account
    </h2>
    <DisplayError
      error={Object {}}
    />
    <label
      htmlFor="email"
    >
      Email
      <input
        name="email"
        onChange={[Function]}
        placeholder="Enter your email"
        type="email"
        value=""
      />
    </label>
    <label
      htmlFor="name"
    >
      Name
      <input
        name="name"
        onChange={[Function]}
        placeholder="Enter your name"
        type="text"
        value=""
      />
    </label>
    <label
      htmlFor="password"
    >
      Password
      <input
        name="password"
        onChange={[Function]}
        placeholder="Enter a password"
        type="password"
        value=""
      />
    </label>
    <button
      type="submit"
    >
      Sign up!
    </button>
  </fieldset>
</form>
`;
